pkg/gadgets/profile/cpu/types/types.go

```diff
@@ -32,3 +32,9 @@ type Report struct {
 	KernelStack []string `json:"kernel_stack,omitempty"`
 	Count       uint64   `json:"count,omitempty"`
 }
+
+// GetBaseEvent is defined to implement commonutils.BaseElement interface
+// required to use commonutils.BaseParser.
+func (e Report) GetBaseEvent() *eventtypes.Event {
+	return nil
```

Comment:
hmm, this is weird. Probably report should contain `eventtypes.Event`.

Reply:
Mmmm I don't think `Report` should contain `eventtypes.Event` because:

1. The profile cpu gadget is not an event-based gadget so it would be misleading. 
2. We should always initialize the `event.Type` as `NORMAL` so that it passes [this check](https://github.com/kinvolk/inspektor-gadget/blob/a6db7da168dfa7c4c47b9adac9f8459d731ecf01/cmd/common/utils/parser.go#L109). And, in my opinion, it could create confusion because, again, it is not an event-based gadget. 
3. If we make this change for `Report`, should we also add `eventtypes.Event` to [Container](https://github.com/kinvolk/inspektor-gadget/blob/a6db7da168dfa7c4c47b9adac9f8459d731ecf01/pkg/container-collection/containers.go#L77-L81)? And to the `Stats` of all the top gadgets ([#782](https://github.com/kinvolk/inspektor-gadget/pull/782) where we start using `BaseParser`). They all use `BaseParser` but don't support special events.

I know it is not the cleanest solution, but I think it could be ok with better comments. Check this:

```golang
// GetBaseEvent is defined to implement the commonutils.BaseElement interface so
// that we can use the commonutils.BaseParser methods to parse Report. In
// commonutils.BaseParser.Transform(), we call GetBaseEvent() to check whether
// the element being parsed is a special event or not. Given that the profile
// cpu gadget is not event-based, this method simply returns nil.
func (e Report) GetBaseEvent() *eventtypes.Event {
	return nil
}
```

```golang
// GetBaseEvent is defined to implement the commonutils.BaseElement interface so
// that we can use the commonutils.BaseParser methods to parse Container. In
// commonutils.BaseParser.Transform(), we call GetBaseEvent() to check whether
// the element being parsed is a special event or not. Given that Container is
// not an event, this method simply returns nil.
func (e Container) GetBaseEvent() *eventtypes.Event {
	return nil
}
```

WDYT?

Reply:
TL;DR: I think we're messing up with `eventtypes.Event` and we need to fix it.

`eventtypes.Event` is used to carry **messages** from the tracer to the client. Each time a gadget produces output it's marshaled into a structure that contains this type, then it's unmarshaled on the client and according to the kind of gadget it could be printed to the screen (trace, snapshot), or saved for future processing (top). I think the `Event` name is misleading, as this is actually used for gadgets that are not event based, like snapshot. 

I see that `eventtypes.Event` is like an "message header" that indicates the kind of event and some debugging information. I'm thinking to rework this in a different way:

1. Rename Event to `DataHeader` and remove `CommonData` from it, leaving only the type and message.
2. Modify types of different gadgets to embed this structure within their types. 
   - Tracers on their Event types (as currently done, they'll have to add CommonData since it's not on eventtypes.Event anymore)
   - Toppers on their "[Event](https://github.com/kinvolk/inspektor-gadget/blob/120ffd1b1ff12012ab7ca1ee6f93b2fea18843de/pkg/gadgets/top/file/types/types.go#L75)" (to be renamed?) type too
   - Profilers on their [Report](https://github.com/kinvolk/inspektor-gadget/blob/120ffd1b1ff12012ab7ca1ee6f93b2fea18843de/pkg/gadgets/profile/cpu/types/types.go#L26) type
   - I think snapshotters should define a new type that wraps the current event:
```golang
type Event struct {
  eventtypes.DataHeader

  Processes/Sockets []Process/Socket
}

type Process/Socket struct {̣ <-- current Event type
```

3. We can know handle the special event in the same way in all gadgets. (Their "Event" have the same header).

Reply:
I like your proposal! It would make things clearer. I will pause this PR until then.

Reply:
@mauriciovasquezbernal there already exists an issue for this? If it doesn't, I will create one, move your proposal there and close this PR.

Reply:
I don't think we have an issue for it.